consensus/ethash/consensus.go

```diff
@@ -286,13 +287,12 @@ func (ethash *Ethash) verifyHeader(chain consensus.ChainReader, header, parent *
 // CalcDifficulty is the difficulty adjustment algorithm. It returns
 // the difficulty that a new block should have when created at time
 // given the parent block's time and difficulty.
-// TODO (karalabe): Move the chain maker into this package and make this private!
-func CalcDifficulty(config *params.ChainConfig, time uint64, parent *types.Header) *big.Int {
+func (ethash *Ethash) CalcDifficulty(chain consensus.ChainReader, time uint64, parent *types.Header) *big.Int {
```

Comment:
If we completely move `CalcDifficulty` inside the ethash engine, then outside code that wants to work with ethash will always need an engine. I think the best would be to leave this method as a plain global, and **also** to wrap it within the engine. That way we have an engine agnostic way to calculate a difficulty, but it can be done without requiring an engine object too.